text/0249-v2-experiments.md

```diff
@@ -374,6 +380,257 @@ Disadvantages:
+  remove the prefix/suffix.
+- A lot of deprecated code in aws-cdk-lib, possibly blowing up `aws-cdk-lib`.
+  This might not be a real concern as we can reuse a lot the code between
+  different version of an API.
+
+**How can we encourage users to upgrade to a newer version?**
+
+When a new minor version introduces a breaking change to an experimental API,
+users have a clear motivation to upgrade to the new version of the API - they
+must do so in order to upgrade to a newer version of the AWS CDK. If the API is
+only deprecated, users have no motivation to upgrade to a newer version of the
+API, even worse, they might not be aware that a newer version exists. One way to
+encourage users to upgrade to a newer version of an API, is to supply tools that
+will inform users. For example, we can add a capability to the `cdk doctor`
+command that will notify users that their CDK application is using deprecated
+experimental APIs.
+
+Executing `cdk doctor` will print:
+
+```bash
+neta@dev/my-cdk-application$ cdk doctor
+
+Newer version of experimental APIs your application is using are available. You should consider upgrading.
+To see which experimental APIs can be upgraded, execute `cdk --show-deprecated-api-usage`
+```
+
+**If there are no breaking changes, are the APIs really experimental?**
+
+Given that experimental APIs are safe to use, and no breaking changes will be
+introduced, we might consider not referring to them as experimental, and simply
+add versioning in the API name. The first version of `grantWrite` will be named
+`grantWriteV1`, the second version will be named `grantWriteV2` and so on. While
+the API will not break, we should still add `experimental` for the following
+reasons:
+
+1. **Real-estate**: `grantWrite` is a much better name than `grantWriteV3` for
+   the final API.
+2. **Encourage feedback**: Declaring an API as experimental encourage the
+   community to supply feedback.
+3. **Setting the right expectations**: When an API is marked as experimental
+   users are aware that this is not the final version of the API, and its
+   deprecation is expected.
+
+### Open questions
+
+If we decide to implement option 6, we will need to formalize the experimental
+naming strategy.
+
+- Should experimental modules be named with `experimental`? e.g `aws-synthetics`
+  be named `aws-synthetics-pre-v1`, or is it sufficient that its types be named
+  as experimental, e.g `CanaryPreV1`.
+- If a type that contains other types is experimental, is it sufficient that the
+  containing type is named as `experimental` or should its properties/methods
+  should be named as experimental as well. e.g if the `Canary` construct is
+  experimental, named `CanaryPreV1`, should its method `createAlarm` be named
+  `createAlarmPreV1`:
+
+  ```ts
+  const canary = new CanaryPreV1(this, 'MyCanary', {
+  ...
+  });
+  canary.createAlarmPreV1(...)
+  // or
+  canary.createAlarm(...)
+  ```
+
```

Comment:
My general feedback on Option 6 is that it's not baked enough yet.

It gives an example of evolving a method, but the interesting part here are the questions that have not been answered:

1. How are modules organized? I.e., I create a new L2. What module does it live in? `aws-cdk-lib`? If so, what's the name of the namespace for that module? If not `aws-cdk-lib`, where does it go?
2. Continuing the example from above ("I create a new L2"): how are the classes and interfaces named? What's the name of the resource interface (i.e., `IBucket`). What's the name of the Construct class with the new L2? What about other, non-Construct classes introduced as part of that L2?

etc.

Right now, this is simply a convention ("I'm not sure about `someMethod()`, let me call it `someMethodVer1()` to hedge my bets"), not a mechanism for doing experiments.

Reply:
Yup. All in `aws-cdk-lib`. 

It is a convention, we need to formalize it for all types. 

If by "a mechanism for doing experiment" you mean a mechanism that allows us to release breaking changes, this PR suggest that any such solution is not viable.

Reply:
I agree with Neta. The goal of this RFC is to allow us to release APIs that are not "final" so that we can get feedback from real use before we commit to the final API. Our analysis shows that there is no viable way to do that and continue to respect semantic versioning. Option 6 proposes viable a model that allows us to achieve this goal.

It is also the exact approach taken by Kubernetes to achieve the same goal, which validates this approach further.

Reply:
> If by "a mechanism for doing experiment" you mean a mechanism that allows us to release breaking changes, this PR suggest that any such solution is not viable.

Not possible to release breaking changes _in `aws-cdk-lib`_.

If we have a separate module with breakable code, like the other 5 options beside option 5 specify, released with version `0.x`, then it definitely _is_ viable (see options 2. and 3. for details).

Reply:
> Yup. All in `aws-cdk-lib`.

Great. Which namespace? Let's say I'm adding the first L2 for a module `aws-xyz` that's currently `cnf-only`. What namespace does the new L2 land in? `aws-cdk-lib/aws-xyz`? Or somewhere else?

> It is a convention, we need to formalize it for all types.

Yes. That's kind of the purpose of this RFC - to propose a solution that covers all of the cases we know we need to deal with. It's hard to judge the merits of option 6 without seeing the entire proposal - how does it treat constructs, props, classes, interface properties, interface methods (will we have `bindVer1()` now, for example?), etc.

Right now, the proposal is a convention we already use, in a slightly different way. For example, in the SecretsManger module, which is stable, we screwed up the `fromSecretName()` method; so now, we have
`fromSecretNameV2()`.

Reply:
And one more aspect that this proposal has to include: what happens with all of the Construct Library modules that are `Experimental` or `DevPreview` in version `1.x` today?

Reply:
I am not sure we need to discuss each api in this RFC, but the general approach for moving from v1 to v2 should be included.

I think that once this decision is made, we need to create concrete plan for implementing it as part of v2. We will need to make an api by api decision for what to do.

Reply:
We will discuss the strategy for migrating dev preview/experimental API from v1 to v2 in a separate doc, after we agree on the mechanism for adding non final APIs.

Reply:
> @skinny85 I was under the impression we are going to flip all v1 modules/APIs to stable in v2 in order to minimize the migration pain for customers, is that not the case?

No, that is not the case. If we adopted a different option than 6, all of the modules that were not "Stable" or "CFN-only" at the time of releasing a GA version of `aws-cdk-lib` version `2.x` would be placed into that solution's designated spot for code that is allowed to be broken (for example, that would be `aws-cdk-lib-experiments` if we chose option 2).